src/main/java/cms/gov/madie/measure/services/MeasureService.java

```diff
@@ -589,4 +591,105 @@ private void updateOrganizationName(Measure measure) {
               });
     }
   }
+
+  public MeasureSet associateCmsId(String username, String qiCoreMeasureId, String qdmMeasureId) {
+    if (qiCoreMeasureId == null || qdmMeasureId == null) {
+      log.info(
+          "CMS ID could not be associated. Measure Ids [{}],[{}} cannot be null",
+          qiCoreMeasureId,
+          qdmMeasureId);
+      throw new InvalidIdException("CMS ID could not be associated. Please try again.");
+    }
+
+    Measure qiCoreMeasure = findMeasureById(qiCoreMeasureId);
+    Measure qdmMeasure = findMeasureById(qdmMeasureId);
+
+    if (qiCoreMeasure == null || qdmMeasure == null) {
+      log.info(
+          "CMS ID could not be associated. Measures with given Ids [{}],[{}] are not found",
+          qiCoreMeasureId,
+          qdmMeasureId);
+      throw new ResourceNotFoundException("CMS ID could not be associated. Please try again.");
+    }
+
+    validateCmsIdAssociation(username, qiCoreMeasure, qdmMeasure);
+
+    MeasureSet measureSet = measureSetService.findByMeasureSetId(qiCoreMeasure.getMeasureSetId());
```

Comment:
```suggestion
    MeasureSet measureSet = qiCoreMeasure.getMeasureSet();
```

Reply:
Also, suggest variable name `qiCoreMeasureSet` to reduce context load.

Reply:
thank you, I'll change it